src/bootstrap/src/core/build_steps/compile.rs

```diff
@@ -1830,13 +1830,28 @@ impl Step for Assemble {
             })
             .collect::<HashSet<_>>();
 
+        let link_std_into_rustc_driver = builder.link_std_into_rustc_driver(target_compiler.host);
         let sysroot = builder.sysroot(target_compiler);
         let rustc_libdir = builder.rustc_libdir(target_compiler);
         t!(fs::create_dir_all(&rustc_libdir));
         let src_libdir = builder.sysroot_libdir(build_compiler, host);
         for f in builder.read_dir(&src_libdir) {
             let filename = f.file_name().into_string().unwrap();
-            if (is_dylib(&filename) || is_debug_info(&filename)) && !proc_macros.contains(&filename)
+
+            // For the later stages which gets distributed only copy over the
+            // `rustc_driver` library so we don't end up with an extra copy of `std`.
+            // If we're not statically linking `std` into `rustc_driver`, just copy every library
+            // to ensure `std` is included.
+            // We still need `std` for the initial stage as the bootstrap compiler may not
+            // have the new `rustc_private` linking behavior.
+            let can_be_rustc_dep = filename.starts_with("rustc_driver-")
+                || filename.starts_with("librustc_driver-")
+                || build_compiler.stage == 0
+                || !link_std_into_rustc_driver;
+
+            if can_be_rustc_dep
+                && (is_dylib(&filename) || is_debug_info(&filename))
+                && !proc_macros.contains(&filename)
```

Comment:
This code did also copy `libLLVM` before this change, now that won't happen. But the LLVM library seems to appear in the target sysroot anyway, probably from some other bootstrap step, so hopefully that won't be an issue.

Reply:
It might be better to just exclude `std` then.

Reply:
Where are you seeing it? It seems absent from the dist artifacts:

![image](https://github.com/user-attachments/assets/ffc4e2eb-3f6a-4aa9-a60b-ae4c8907f61d)

Reply:
I pushed a commit that does so.